.github/workflows/main.yml

```diff
@@ -0,0 +1,209 @@
+        uses: actions/checkout@v4
+
+      - name: Setup Environment
+        uses: ./.github/actions/setup
+        with:
+          clippy: true
+          rustfmt: true
+          cargo-cache-key: cargo-interface-lint
+          cargo-cache-fallback-key: cargo-interface
+
+      - name: Format
+        run: pnpm interface:format
+
+      - name: Lint
+        run: pnpm interface:lint
+
+  format_and_lint_programs:
+    name: Format & Lint Programs
+    runs-on: ubuntu-latest
+    steps:
+      - name: Git Checkout
+        uses: actions/checkout@v4
+
+      - name: Setup Environment
+        uses: ./.github/actions/setup
+        with:
+          clippy: true
+          rustfmt: true
+          cli: true
+          cargo-cache-key: cargo-programs-lint
+          cargo-cache-fallback-key: cargo-programs
+
+      - name: Format
+        run: pnpm programs:format
+
+      - name: Lint
+        run: pnpm programs:lint
+
+  format_and_lint_client_js_legacy:
+    name: Format & Lint Client JS Legacy
+    runs-on: ubuntu-latest
+    steps:
+      - name: Git Checkout
+        uses: actions/checkout@v4
+
+      - name: Setup Environment
+        uses: ./.github/actions/setup
+
+      - name: Format Client JS Legacy
+        run: pnpm clients:js-legacy:format
+
+      - name: Lint Client JS Legacy
+        run: pnpm clients:js-legacy:lint
+
+  audit_rust:
+    name: Audit Rust
+    runs-on: ubuntu-latest
+    steps:
+      - name: Git Checkout
+        uses: actions/checkout@v4
+
+      - name: Setup Environment
+        uses: ./.github/actions/setup
+        with:
+          cargo-cache-key: cargo-audit
+
+      - name: Install cargo-audit
+        uses: taiki-e/install-action@v2
+        with:
+          tool: cargo-audit
+
+      - name: Run cargo-audit
+        run: pnpm rust:audit
+
+  semver_rust:
+    if: false # re-enable once spl-pod v0.5.1 is released
+    name: Check semver Rust
+    runs-on: ubuntu-latest
+    steps:
+      - name: Git Checkout
+        uses: actions/checkout@v4
+
+      - name: Setup Environment
+        uses: ./.github/actions/setup
+        with:
+          cli: true
+          cargo-cache-key: cargo-semver
+
+      - name: Install cargo-semver-checks
+        uses: taiki-e/install-action@v2
+        with:
+          tool: cargo-semver-checks
+
+      - name: Run semver checks
+        run: pnpm rust:semver
+
+  spellcheck_rust:
+    name: Spellcheck Rust
+    runs-on: ubuntu-latest
+    steps:
+      - name: Git Checkout
+        uses: actions/checkout@v4
+
+      - name: Setup Environment
+        uses: ./.github/actions/setup
+        with:
+          cargo-cache-key: cargo-spellcheck
+
+      - name: Install cargo-spellcheck
+        uses: taiki-e/install-action@v2
+        with:
+          tool: cargo-spellcheck
+
+      - name: Run cargo-spellcheck
+        run: pnpm rust:spellcheck
+
+  build_programs:
+    name: Build programs
+    runs-on: ubuntu-latest
+    needs: format_and_lint_programs
```

Comment:
Comment for discussion: on the stake repository we recently decoupled the "build" from the "format_and_lint" so they run in parallel.